packages/integrations/src/offline.ts

```diff
@@ -0,0 +1,101 @@
+import { captureEvent } from '@sentry/minimal';
+import { Event, EventProcessor, Hub, Integration } from '@sentry/types';
+import { getGlobalObject, logger, uuid4 } from '@sentry/utils';
+// @ts-ignore
+import localforage = require('localforage');
+
+/**
+ * cache offline errors and send when connected
+ */
+export class Offline implements Integration {
+  // tslint:disable: no-unsafe-any
+
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'Offline';
+
+  /**
+   * @inheritDoc
+   */
+  public readonly name: string = Offline.id;
+
+  /**
+   * event cache
+   */
+  public offlineEventStore: any;
+
+  /**
+   * @inheritDoc
+   */
+  public constructor() {
+    this.offlineEventStore = localforage.createInstance({
```

Comment:
Could `createInstance` fail? What happens to `this.offlineEventStore` then?

We can import localforage typings if they exist as well.

Also, what browsers does `localforage` support? It's fine if it doesn't match Sentry's (https://docs.sentry.io/platforms/javascript/#browser-table) - we just have to document it then.

Reply:
Ah, good catch. I meant to wrap this in a `try/catch`. I will look into importing the type file too.

Reply:
I think it meets the compatibility requirements. It is a wrapper around multiple storage mechanisms. It attempts to use IndexedDB, then falls back to WebSQL, and if all else fails, it falls back to localStorage.
https://github.com/localForage/localForage/wiki/Supported-Browsers-Platforms

Reply:
Nice ok, makes sense. Just one last thing, what happens when local storage is full? Is it a queue, or do new events just get dropped?

@bruno-garcia, how do we handle offline queue in mobile right now?

Reply:
Only envelopes are cached. The order they were cached is kept when attempting to send things to Sentry (i.e device is back online). If `maxItems` is reached (by default 30 IIRC), the oldest file is deleted to make space to the newest one.

We don't use dependencies or a DB of any sort. It's just files written to a directory.

If session data is stored (release health), the `init=true` flag must be preserved. So when deleting a file to make room for a new one (max items reached), the envelope would need to be unwrapped, checked for sessions and if init=true exists, that must be moved to the next session update queued up for submission (from oldest to newest).
Events capture when session tracking is on must be in an envelope together with a session update.

Note that the order being kept means that if you get a call to `captureException`, you won't be sending that to Sentry if there's stuff cached. You must prioritize what's on the storage first.

Reply:
Thanks Bruno for coming through 🚀. We can ignore all the sessions stuff, and envelope details - although important for me, probably not as for you @voraciousdev

Basically there are two things we need to keep in mind.
1) Enforce some kind of `maxItems` in local storage. This can be changed through an option.

2) Make sure that we send all events from local storage first before we continue on with other events. I know we are using an event listener right now to listen to `online`, do you think that is good enough? Maybe we need another check in the `globalEventProcessor` to be sure.

As for how this will work with sessions, we can get to that when release health comes to JS 😄

Reply:
@AbhiPrasad the `setItem` promise will be rejected if there is not enough space on the device. With the way it's currently written in this integration, that would result in `logger.warn('could not cache event while offline')` being called.

Reply:
Ok perfect we should we good then.

Reply:
> Make sure that we send all events from local storage first before we continue on with other events. I know we are using an event listener right now to listen to `online`, do you think that is good enough? Maybe we need another check in the `globalEventProcessor` to be sure.

@AbhiPrasad I'm not sure if this is directed at me, but there is definitely a possible race condition here regarding the `online` event. Given the nature of this integration (to support offline-capable apps), I think it's a real scenario that we could have real error events triggered as soon as the `online` event is fired. For offline apps, it's common practice to listen to the `online` event and run some code (which could then lead to an error being thrown/recorded).

Reply:
Just added localforage typings as well. Good call on that. 👍